kernel/typeops.ml

```diff
@@ -151,27 +151,33 @@ let make_judgev c t =
   Array.map2 make_judge c t
 
 let type_of_apply env func funt argsv argstv =
+  let open CClosure in
   let len = Array.length argsv in
-  let rec apply_rec i typ = 
-    if Int.equal i len then typ
-    else 
-      (match kind (whd_all env typ) with
-      | Prod (_,c1,c2) ->
-	let arg = argsv.(i) and argt = argstv.(i) in
-	  (try
-	     let () = conv_leq false env argt c1 in
-	       apply_rec (i+1) (subst1 arg c2)
-	   with NotConvertible ->
-	     error_cant_apply_bad_type env
-	       (i+1,c1,argt)
-	       (make_judge func funt)
-	       (make_judgev argsv argstv))
-	    
+  let infos = create_clos_infos all env in
+  let tab = create_tab () in
+  let rec apply_rec i typ =
+    if Int.equal i len then term_of_fconstr typ
+    else
+      let typ, _ = whd_stack infos tab typ [] in
+      match fterm_of typ with
+      | FProd (na, c1, c2, e) ->
+        let arg = argsv.(i) in
+        let argt = argstv.(i) in
+        let c1 = term_of_fconstr (mk_clos e c1) in
+        begin match conv_leq false env argt c1 with
```

Comment:
I agree that the fast path is important. I'm not well versed enough to know how expensive the conversion check is in the case of syntactic equality, but I'll keep looking at the code. Suppose that I have two larger terms that differ only at some lead node. Syntactic equality will fail, but conversion will succeed. Will the conversion be slow (because it will reduce the full term) or fast because it will only convert the small term?